<img src="images/bannerUgentDwengo.png" alt="BannerUGentDwengo" style="width:250px;"/>

<div style='color: #690027;' markdown="1">
    <h1>REGELGEBASEERDE SENTIMENTANALYSE</h1> 
</div>

<div class="alert alert-box alert-success">
Taaltechnologen doen een beroep op machine learning-modellen om bij gegeven teksten onderzoek te doen naar sentimentwoorden. In deze notebook maak je kennis met de principes van hun onderzoek. <br> Het gebruik van technologie wordt steeds toegankelijker voor werknemers in een niet-technologische sector, zoals taalkundigen, communicatiewetenschappers, historici en juristen. <br>Dankzij de zeer toegankelijke programmeertaal Python zal ook jij enkele mogelijkheden van taaltechnologie ontdekken.
</div>

<div class="alert alert-box alert-warning">
Als voorbereiding op deze notebook verdiep je je best in de notebooks 'Strings', 'Lists' en 'Dictionaries'. Je maakt best ook kennis met enkele programmeerstructuren in de notebook 'Structuren. Toepassingen bij strings, lists en dictionaries'. 
</div>

<div style='color: #690027;' markdown="1">
    <h2>1. Principes van regelgebaseerde sentimentanalyse</h2> 
</div>

Voor **regelgebaseerde** sentimentanalyse maak je gebruik van een (bestaand) **lexicon** met daarin woorden gekoppeld aan hun **polariteit** (positief, negatief of neutraal), dus een woordenboek van sentimentwoorden. 

'Blij' bijvoorbeeld heeft een positieve polariteit, 'thuisbankieren' een neutrale en 'boos' een negatieve polariteit. In het lexicon wordt de polariteit weergegeven door een reëel getal tussen -2 en 2. Een strikt positief getal komt overeen met een positief sentiment, een strikt negatief getal met een negatief sentiment en 0 met een neutraal sentiment. 

<div>
<img src="images/schaal.png" alt="Banner" align="center" style="width:500px;"/>
</div>

De polariteit van een tekst wordt gegeven door de som van de polariteiten van de sentimentwoorden in die tekst.

Voor je sentimentwoorden uit een lexicon kunt matchen met de gegeven tekst (de data) moet je de data 
-  1) *inlezen*; 
-  2) *preprocessen*, d.w.z. voorverwerken voor *lexicon matching* of voor *machine learning*. 

Preprocessing omvat alle stappen die nodig zijn om de data voor te bereiden op wat volgt, of dat nu een eenvoudige lexicon matching is, dan wel een ingewikkeld machine learning-systeem dat op de data getraind zal worden. <br>Hieronder vind je een oplijsting van veelvoorkomende preprocessing stappen.


#### Preprocessing

* **Lowercasing:** alle hoofdlettertekens worden vervangen door kleine letters. Lowercasing is nodig omdat de woorden in een lexicon staan zonder hoofdletters.
* **Tokenisering:** alle zinnen worden in betekenisvolle eenheden of 'tokens' gesplitst, zoals woorden en leestekens. Deze splitsing gebeurt op basis van de aanwezige spaties in de zinnen; daarom zullen de woorden van elkaar moeten gescheiden zijn door een spatie.  
* **Part-of-speech tagging:** aan elk token wordt de grammaticale woordcategorie toegekend, zoals adjectief of symbool. Sommige woorden kunnen bv. als zelfstandig én als bijvoeglijk naamwoord voorkomen. Zo'n woord kan ook een andere sentimentwaarde hebben naargelang zijn woordsoort.
* **Lemmatisering:** alle tokens worden omgezet naar hun lemma of woordenboekvorm (bv. een substantief komt in een woordenboek in het enkelvoud voor en van een werkwoord vind je er de infinitief). Die woordenboekvorm wordt dan opgezocht in het lexicon. 

#### Voorbeeld:
Gegeven zin:    
-  De spelletjes waren toffe ijsbrekers.

Lowercasing:    
-  de spelletjes waren toffe ijsbrekers.

Tokens:         
-  'de' 'spelletjes' 'waren' 'toffe' 'ijsbrekers' '.' 

Part-of-speech: 
-  'de': lidwoord;
-  'spelletjes': substantief; 
-  'waren': werkwoord;
-  'toffe': adjectief;
-  'ijsbrekers': substantief;
-  '.': leesteken (symbool).

Lemma's: 'de', 'spel', 'zijn', 'tof', 'ijsbreker', '.'

Polariteit:
-  De polariteiten van de lemma's worden opgezocht in het lexicon; lidwoorden en leestekens zijn daarbij niet van belang.  
-  'spel' heeft polariteit 1, 'zijn' heeft polariteit 0, 'tof' heeft polariteit 0,8 en 'ijsbreker' 0.
-  De polariteit van de gegeven zin is de som van deze polariteiten, dus 1,8. 
-  1,8 is een positief getal. De zin roept een positief sentiment op. 

<div style='color: #690027;' markdown="1">
    <h2>2. Het lexicon</h2> 
</div>

### Modules importeren 

Nu je dit weet, kun je bijna aan de slag. Je laadt eerst twee Python-modules in. <br>
Voer daartoe de code-cel hieronder uit. 

<div class="alert alert-block alert-info"> 
Python is vaak zeer intuïtief in gebruik en bovendien zo populair dat er heel wat modules voorhanden zijn die men vrij kan gebruiken. In een module zitten heel wat functies vervat die ervaren informatici reeds voor jou hebben geprogrammeerd. 
</div>

In [1]:
# modules importeren
import pickle                     # voor lexicon
from colorama import Fore, Back   # om in kleur te kunnen printen
import string                     # voor opsomming leestekens 
from lexiconhulp import tienelementen   

### Lexicon inlezen 

Voer de code-cel hieronder uit. De code in deze cel hoef je niet in detail te begrijpen.

In [2]:
# lexicon inlezen 
with open("data/new_lexicondict.pickle", "rb") as file: # bestand lexicondict.pickle in map data bevat het sentimentlexicon
    lexicon = pickle.load(file)

In [3]:
type(lexicon)

dict

In [4]:
# aantal elementen in lexicon
len(lexicon)

10938

In [5]:
# toon tienelementen van lexicon
print(tienelementen(lexicon))

'retorisch': {'postag': ['ADJ'], 'polarity': [0.0]}
'gezwind': {'postag': ['ADJ'], 'polarity': [0.6]}
'evenwichtig': {'postag': ['ADJ'], 'polarity': [1.25]}
'modaal': {'postag': ['ADJ'], 'polarity': [0.4]}
'digitaal': {'postag': ['ADJ'], 'polarity': [0.0]}
'fout': {'postag': ['ADJ', 'NOUN'], 'polarity': [-0.5, -2.0]}
'onverdeeld': {'postag': ['ADJ'], 'polarity': [0.1]}
'wulps': {'postag': ['ADJ'], 'polarity': [0.7]}
'bemoeiziek': {'postag': ['ADJ'], 'polarity': [-1.35]}



<div class="alert alert-block alert-info"> 
    Het lexicon is een <b>dictionary</b> met 10 938 woorden.  Het lexicon geeft de woordsoort (part-of-speech tag, 'postag') en de polariteit ('polarity') van de woorden in het lexicon.<br>
    De woorden in het lexicon zijn de <b>keys</b> van de dictionary.
    De <b>values</b> van deze dictionary zijn zelf een dictionary met twee keys ("postag" en "polarity") die beide een list als value hebben, een list met hoogstens 2 elementen.
</div>

Enkele woorden uit het lexicon:<br><br>  'retorisch': {'postag': ['ADJ'], 'polarity': [0.0]}, <br> 'gezwind': {'postag': ['ADJ'], 'polarity': [0.6]}, <br>'evenwichtig': {'postag': ['ADJ'], 'polarity': [1.25]},<br> 'modaal': {'postag': ['ADJ'], 'polarity': [0.4]},<br> 'digitaal': {'postag': ['ADJ'], 'polarity': [0.0]}, <br>'fout': {'postag': ['ADJ', 'NOUN'], 'polarity': [-0.5, -2.0]}<br>'analfabeet': {'postag': ['NOUN'], 'polarity': [-1.0]}<br>'stigmatiseren': {'postag': ['VERB'], 'polarity': [-2.0]}

Hieruit leid je bv. af:
-  het woord 'retorisch' is een adjectief dat qua sentiment een neutrale polariteit heeft;
-  het woord 'gezwind' is een adjectief dat qua sentiment een positieve polariteit heeft;
-  het woord 'evenwichtig' is een adjectief dat qua sentiment ook een positieve polariteit heeft, maar het wordt positiever aangevoeld dan 'gezwind';
-  het woord 'fout' kan zowel een adjectief als een substantief zijn, beide met een negatieve polariteit, maar als substantief wordt het als meer negatief aangevoeld dan als adjectief;
- het woord 'analfabeet' is een substantief dat qua sentiment een negatieve polariteit heeft;
- het woord 'stigmatiseren' is een werkwoord dat qua sentiment een negatieve polariteit heeft, meer negatief dan 'analfabeet'.

Je zou het lexicon dus ook kunnen weergeven in de vorm van een tabel:<br><br>

<table>
 <thead align="center">
    <tr>
      <td>woord</td>
      <td>postag</td>
      <td>polarity</td>
     </tr>    
  </thead>
  <tbody align="center">  
      <tr> <td> retorisch </td>   <td> ADJ </td> <td> 0.0 </td>  </tr> 
      <tr> <td> gezwind </td>     <td> ADJ </td> <td> 0.6 </td>  </tr> 
      <tr> <td> evenwichtig </td> <td> ADJ </td> <td> 1.25 </td>  </tr> 
      <tr> <td> modaal </td>      <td> ADJ </td> <td> 0.3 </td>  </tr> 
      <tr> <td> digitaal </td>    <td> ADJ </td> <td> 0.0 </td> </tr> 
      <tr> <td> fout </td>        <td> ADJ </td> <td> -0.5 </td> </tr> 
      <tr> <td> fout </td>        <td> NOUN </td> <td> -2.0 </td> </tr> 
    </tbody>           
</table>

**Merk op** dat de woordsoorten in het Engels teruggegeven worden. 

-  *'NOUN'* staat voor een zelfstandig naamwoord of substantief; 
-  *'ADJ'* voor een bijvoeglijk naamwoord of adjectief; 
-  *'ADV'* voor een bijwoord; 
-  *'DET'* voor een lidwoord; 
-  *'VERB'* voor een werkwoord; 
-  *'AUX' voor een hulpwerkwoord;
-  *'PRON'* voor een voornaamwoord;
-  *'PROPN'* voor een eigennaam, enz. 
-  Aan leestekens wordt *'SYM'* toegekend.

Je kan aan de hand van Python-code de woordsoort en de polariteit van een woord gemakkelijk opzoeken in het lexicon:
-  `lexicon["retorisch"]["postag"]` heeft als uitvoer `['ADJ']`
-  `lexicon["retorisch"]["polarity"]` heeft als uitvoer `[0.0]`
-  `lexicon["fout"]["postag"]` heeft als uitvoer `['ADJ', 'NOUN']`
-  `lexicon["fout"]["polarity"]` heeft als uitvoer `[-0.5, -2.0]`

Test dit uit:

In [6]:
lexicon["retorisch"]["postag"]

['ADJ']

In [7]:
lexicon["retorisch"]["polarity"]

[0.0]

In [8]:
lexicon["fout"]["postag"]

['ADJ', 'NOUN']

In [9]:
lexicon["fout"]["polarity"]

[-0.5, -2.0]

#### Oefening 2.1:

Zoek op in het lexicon:
-  de woordsoort van 'kwekkebekken'

In [10]:
#voorbeeldscript
lexicon["kwekkebekken"]["postag"]

['VERB']

-  de woordsoort van  'aanraden'

In [11]:
#voorbeeldscript
lexicon["aanraden"]["postag"]

['ADJ', 'VERB']

-  de polariteit van 'jolijt'

In [12]:
#voorbeeldscript
lexicon["jolijt"]["polarity"]

[2.0]

-  de polariteit van 'konkelen'

In [13]:
#voorbeeldscript
lexicon["konkelen"]["polarity"]

[-1.0]

De woorden die in het lexicon staan, noemt men *sleutels* of *keys* in Python. Je kan ze opvragen via de volgende code-cel. 

In [14]:
lexicon.keys()

dict_keys(['retorisch', 'gezwind', 'evenwichtig', 'modaal', 'digitaal', 'fout', 'onverdeeld', 'wulps', 'bemoeiziek', 'grootscheeps', 'vorig', 'creatief', 'tegenoverliggend', 'schalks', 'kurkdroog', 'kwistig', 'verknipt', 'tiranniek', 'mondig', 'deels', 'verrassend', 'tomeloos', 'schadelijk', 'luguber', 'protestants', 'uiterst', 'eerloos', 'loodgrijs', 'zogeheten', 'harmonieus', 'fijnbesnaard', 'plechtig', 'gemaakt', 'ondubbelzinnig', 'overijverig', 'psychisch', 'vast', 'schoon', 'oubollig', 'basaal', 'filantropisch', 'steengoed', 'kouwelijk', 'sefardisch', 'tussenliggend', 'zelfvoldaan', 'geheimzinnig', 'impulsief', 'definitief', 'aandoenlijk', 'storend', 'plechtstatig', 'koortsachtig', 'doortimmerd', 'humeurig', 'doodgeboren', 'bekakt', 'categoriaal', 'consistent', 'fit', 'spoedig', 'direct', 'overspannen', 'zuurverdiend', 'ongelezen', 'onbewogen', 'fortuinlijk', 'verkeken', 'pikant', 'onhebbelijk', 'overdadig', 'heimelijk', 'krengerig', 'milieubewust', 'dakloos', 'consequent', 'blut'

Zo, je bent klaar voor een toepassing. Stap 1: de data (gegevens) inlezen en bekijken. 

<div style='color: #690027;' markdown="1">
    <h2>3. Toepassing: klantenreview</h2> 
</div>

In wat volgt zal je sentimentanalyse uitvoeren op een gegeven review. 

### De data 

Voer de volgende code-cel uit om de review in te lezen en vervolgens te bekijken.

In [15]:
review = "Nieuw concept in Gent, maar dat kan volgens mij toch beter. De meeste cornflakes waren gewoon de basic soorten. Ook wat duur voor de hoeveelheid die je krijgt, vooral met de toppings zijn ze zuinig. En als je ontbijt aanbiedt, geef de mensen dan toch ook wat meer keuze voor hun koffie."

In [16]:
print(review)

Nieuw concept in Gent, maar dat kan volgens mij toch beter. De meeste cornflakes waren gewoon de basic soorten. Ook wat duur voor de hoeveelheid die je krijgt, vooral met de toppings zijn ze zuinig. En als je ontbijt aanbiedt, geef de mensen dan toch ook wat meer keuze voor hun koffie.


Je bent klaar voor stap 2: de preprocessing uitvoeren op de review.

<div style='color: #690027;' markdown="1">
    <h2>4. Preprocessing</h2> 
</div>

### Lowercasing
In deze stap zet je de tekst in de review om naar kleine letters. Lowercasing is nodig omdat woorden zonder hoofdletter in het lexicon staan.

De variabele `review_kleineletters` verwijst naar deze omgezette tekst.

In [17]:
# zet tekst van de review om naar tekst in kleine letters
review_kleineletters = review.lower()  # review met kleine letters schrijven

In [18]:
# toon resultaat van lowercasing
print(review_kleineletters)

nieuw concept in gent, maar dat kan volgens mij toch beter. de meeste cornflakes waren gewoon de basic soorten. ook wat duur voor de hoeveelheid die je krijgt, vooral met de toppings zijn ze zuinig. en als je ontbijt aanbiedt, geef de mensen dan toch ook wat meer keuze voor hun koffie.


### Tokenisering
Nu zal je de review in woorden en leestekens opsplitsen m.b.v. de computer, dat gebeurt op basis van spaties.
Deze woorden en leestekens zijn **tokens**.

Om de tokens automatisch te laten genereren, is het nodig dat de woorden en leestekens van elkaar gescheiden zijn door een spatie. Na elke spatie kan dan een nieuwe token worden gegenereerd.
Bv. Hello, world!   wordt eerst geschreven als Hello , world ! en de vier tokens zijn dan: 'Hello', ',', 'world' en '!'.

Je zal de review dus eerst wat moeten aanpassen: voor en na elk leesteken moet zeker een spatie aanwezig zijn.  

#### Spaties invoeren vóór elk leesteken
In deze stap plaats je in de review een spatie voor (en na) elk leesteken.

De variabele `review_spatie` verwijst naar deze omgezette tekst.

In [19]:
leestekens = string.punctuation
print(leestekens)

!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~


In [20]:
# spaties toevoegen aan tekst van review
review_spatie = ""     # lege string
for karakter in review_kleineletters:
    if karakter not in leestekens:
        review_spatie = review_spatie + karakter
    else:
        review_spatie = review_spatie + " " + karakter + " "     # spatie voor en na een leesteken

In [21]:
# toon resultaat van spaties toevoegen
print(review_spatie)

nieuw concept in gent ,  maar dat kan volgens mij toch beter .  de meeste cornflakes waren gewoon de basic soorten .  ook wat duur voor de hoeveelheid die je krijgt ,  vooral met de toppings zijn ze zuinig .  en als je ontbijt aanbiedt ,  geef de mensen dan toch ook wat meer keuze voor hun koffie . 


In [22]:
# tokenisering
tokens = review_spatie.split()  

In [23]:
# toon resultaat van tokenisering
print(tokens)

['nieuw', 'concept', 'in', 'gent', ',', 'maar', 'dat', 'kan', 'volgens', 'mij', 'toch', 'beter', '.', 'de', 'meeste', 'cornflakes', 'waren', 'gewoon', 'de', 'basic', 'soorten', '.', 'ook', 'wat', 'duur', 'voor', 'de', 'hoeveelheid', 'die', 'je', 'krijgt', ',', 'vooral', 'met', 'de', 'toppings', 'zijn', 'ze', 'zuinig', '.', 'en', 'als', 'je', 'ontbijt', 'aanbiedt', ',', 'geef', 'de', 'mensen', 'dan', 'toch', 'ook', 'wat', 'meer', 'keuze', 'voor', 'hun', 'koffie', '.']


Zo krijg je dus een lijst van de tokens.

Als de tekst getokeniseerd is, kan je aan elk token een **part-of-speech tag** toekennen, voor een woord is dat de woordsoort.<br>Bij het opzoeken van een token in het lexicon is het belangrijk dat je daarbij ook de part-of-speech tag controleert. Sommige woorden kunnen immers als substantief én als adjectief gebruikt worden.

Een token opzoeken in het lexicon doe je in zijn woordenboekgedaante. Elk token moet dus gelemmatiseerd worden, m.a.w. teruggebracht worden naar zijn **lemma** of woordenboekvorm, zoals enkelvoud voor een substantief en de infinitief voor een werkwoord. Het opzoeken van zo'n woordenboekvorm in het lexicon kan dan **geautomatiseerd** worden.

### Part-of-speech tagging en lemmatisering

Je wilt dus een lijst van de part-of-speechs tags en van de lemma's die bij de tokens horen, en deze lemma's kan je dan opzoeken in het lexicon.

Het eerste token is 'nieuw'. De basisvorm of lemma is `nieuw` en dit is een adjectief, ADJ.

Neem het token ','. De basisvorm of lemma is `,` en dit is een leesteken, dus een symbool, SYM.

Neem het token 'kan'. De basisvorm of lemma is `kunnen` en dit is een werkwoord, VERB.

Neem het token 'soorten'. De basisvorm of lemma is `soort` en dit is een substantief, NOUN.

**Merk op** dat het lemma van tokens zoals *'waren'*, *'soorten'*, *'krijgt'*, *'toppings'*, *'aanbiedt'*, *'geef'* en *'mensen'* niet hetzelfde is als het token. Bij andere tokens, zoals *'nieuw'*, is dat wel het geval.

Maak lijsten van de lemma's en de part-of-speech tags.<br> 
De volgorde in deze lijsten is van belang. Het eerste element in de ene lijst moet overeenkomen met het eerste element in de andere lijst; het tweede element in de ene lijst moet overeenkomen met het tweede element in de andere lijst; enz. <br>
Tip: vertrek van de lijst van de tokens en gebruik 'copy-paste' (CTRL-C, CTRL-V), pas dan enkel de lemma's aan die verschillen van het token. De postags zal je allemaal moeten intypen.

In [24]:
postags = ['ADJ', 'NOUN', 'ADP', 'PROPN', 'SYM', 'CCONJ', 'PRON', 'VERB', 'ADP', 'PRON', 'ADV', 'ADJ', 'SYM', 'DET', 'ADV', 'NOUN', 'AUX', 'ADJ', 'DET', 'ADJ', 'NOUN', 'SYM', 'ADV', 'DET', 'ADJ', 'ADP', 'DET', 'NOUN', 'PRON', 'PRON', 'AUX', 'SYM', 'ADV', 'ADP', 'DET', 'NOUN', 'AUX', 'PRON', 'ADJ', 'SYM', 'CCONJ', 'SCONJ', 'PRON', 'NOUN', 'AUX', 'SYM', 'VERB', 'DET', 'NOUN', 'ADV', 'ADV', 'ADV', 'PRON', 'DET', 'NOUN', 'ADP', 'PRON', 'NOUN', 'SYM', 'SYM', 'SYM']
lemmas = ['nieuw', 'concept', 'in', 'gent', ',', 'maar', 'dat', 'kunnen', 'volgens', 'mij', 'toch', 'goed', '.', 'de', 'veel', 'cornflakes', 'zijn', 'gewoon', 'de', 'basic', 'soort', '.', 'ook', 'wat', 'duur', 'voor', 'de', 'hoeveelheid', 'die', 'je', 'krijgen', ',', 'vooral', 'met', 'de', 'topping', 'zijn', 'ze', 'zuinig', '.', 'en', 'als', 'je', 'ontbijt', 'aanbieden', ',', 'geven', 'de', 'mens', 'dan', 'toch', 'ook', 'wat', 'meer', 'keuze', 'voor', 'hun', 'koffie', '.', '.', '.']

Het opstellen van deze lijsten kost wel wat tijd aangezien ze grotendeels manueel gebeuren.  

<div style='color: #690027;' markdown="1">
    <h2>5. Sentiment lexicon matching</h2> 
</div>

Nu je review *gepreprocessed (voorverwerkt)* is, kan je het **sentiment bepalen** met behulp van het lexicon van sentimentwoorden dat je ter beschikking hebt.<br>
Ook hier werk je in twee grote stappen: 
-  je zoekt de sentimentwoorden in de review op in het lexicon, m.a.w. je kijkt welke tokens a.d.h.v. hun lemma in het lexicon te vinden zijn (je past 'matching' toe op de review);
-  je hebt de polariteit van de sentimentwoorden nodig volgens hun woordsoort in de review.

#### Voorbeeld 5.1 
Beschouw het lemma `"nieuw"`.
Je zoekt dit lemma op in het sentimentlexicon.

In [25]:
# staat "nieuw" in het lexicon?
"nieuw" in lexicon

True

Het staat er dus in. Vraag de part-of-speech tag ("postag") en de polariteit ("polarity") van `"nieuw"` op.

In [26]:
lexicon["nieuw"]["postag"] 

['ADJ']

In [27]:
lexicon["nieuw"]["polarity"] 

[0.575]

#### Oefening 5.1 
Doe hetzelfde voor het lemma `"duur"`.

In [28]:
# voorbeeldscript
# staat "duur" in het lexicon?
"duur" in lexicon

True

In [29]:
# voorbeeldscript
print(lexicon["duur"]["postag"]) 
print(lexicon["duur"]["polarity"]) 

['ADJ', 'NOUN']
[-1.0666666666666667, -2.0]


#### Voorbeeld 5.2
De index van `"NOUN"` in de lijst is 1.<br>
Met de volgende code vraag je de polariteit op van `"duur"` als zelfstandig naamwoord:

In [30]:
lexicon["duur"]["polarity"][1] 

-2.0

#### Voorbeeld 5.3
-  Voor de sentimentanalyse van de review maak je nu een lijst van de **sentimentwoorden** van de review: de tokens die in het lexicon voorkomen. Naar deze lijst verwijs je met de variabele `lexiconmatches`.<br> 
-  Je maakt ook een lijst met de polariteiten van deze tokens. Naar deze lijst verwijs je met de variabele `polariteiten`. 

Je overloopt daarvoor een voor een alle lemma's. Voor elk lemma dat in het lexicon staat, controleer je de woordsoort, part-of-speech tag. Bij een correcte woordsoort voeg je het overeenkomstige token toe aan de lijst met tokens en de overeenkomstige polariteit aan de lijst met polariteiten. 

Voor `"nieuw"` betekent dit dat `"nieuw"` wordt toegevoegd aan de lijst `lexiconmatches` en `0.575` aan de lijst `polariteiten`.

Tot slot tel je alle polariteiten bij elkaar op. De som, `sum(polariteiten)`, geeft het sentiment van de review. 

In [31]:
# zoek lexicon matches in de review
lexiconmatches = []       # lege lijst, op te vullen met tokens van de lemma's gevonden in lexicon
polariteiten = []         # lege lijst, op te vullen met polariteiten van gevonden tokens 

# beschouw lemma's met overeenkomstige woordsoort en token
for lemma, postag, token in zip(lemmas, postags, tokens):
    if lemma in lexicon.keys() and postag in lexicon[lemma]["postag"]:  
            lexiconmatches.append(token)                      # overeenkomstig token toevoegen aan lijst lexiconmatches
            if postag == lexicon[lemma]["postag"][0]:
                polariteiten.append(lexicon[lemma]["polarity"][0])
            else:
                polariteiten.append(lexicon[lemma]["polarity"][1])
                # overeenkomstige polariteit toevoegen aan lijst polariteiten
    # lemma moet aanwezig zijn in lexicon
    # alleen wanneer het lemma en de POS-tag overeenkomen, is er een match (zie bv. 'fout' als ADJ en 'fout' als NOUN) 

# polariteit review
polariteit = sum(polariteiten)

# eindbeslissing voor deze review
if polariteit > 0:
    sentiment = "positief"
elif polariteit == 0:
    sentiment = "neutraal"
elif polariteit < 0:
    sentiment = "negatief"
print("De polariteit van de review is: " +str(polariteit))
print("Het sentiment van de review is " + sentiment + ".")    

De polariteit van de review is: 4.783333333333333
Het sentiment van de review is positief.


#### Oefening 5.2
Sommige zaken verliepen reeds geautomatiseerd, sommige moest je manueel doen. 
Lijst eens op wat manueel gebeurde en wat automatisch.

Antwoord:

Antwoord: 

Automatisch: kleine letters, spaties rond de leestekens, tokenisering, lexicon matching en polariteiten eruit halen, relevante tokens en polariteiten bewaren, polariteit en sentiment review bepalen.
Manueel: lemmatisering, woordsoort bepalen, lemma's en postags bewaren in lijsten.

<div style='color: #690027;' markdown="1">
    <h2>6. Sentiment lexicon matching: achter de schermen</h2> 
</div>

Bekijk even de (gematchte) sentimentwoorden en hun polariteit door ze op te vragen.

In [32]:
print(lexiconmatches)
print(polariteiten)

['nieuw', 'kan', 'beter', 'gewoon', 'duur', 'voor', 'zuinig', 'geef', 'mensen', 'voor']
[0.575, 1.0, 0.775, 0.5, -1.0666666666666667, 1.0, 0.0, 1.0, 0.0, 1.0]


Je kan in de gegeven review de **sentimentwoorden markeren**: groen voor een positieve polariteit, rood voor een negatieve en blauw voor een neutrale. 

Je vertrekt daarvoor van de oorspronkelijke tekst. In deze tekst vervang je de tokens die sentimentwoorden zijn door zichzelf op een gekleurde achtergrond. De niet-sentimentwoorden laat je ongemoeid.

In [33]:
review_highlighted = review_spatie    # neem review waarin spaties werden aangebracht
         
for token, polariteit in zip(lexiconmatches, polariteiten):
    if polariteit > 0: # overeenkomstige polariteit is positief
        review_highlighted = review_highlighted.replace(token, Back.GREEN + token + Back.RESET)   # positieve token groen markeren 
    elif polariteit == 0.0: # overeenkomstige polariteit is neutraal
        review_highlighted = review_highlighted.replace(token, Back.BLUE + token + Back.RESET)    # neutrale token blauw markeren
    elif polariteit < 0: # negatieve polariteit
        review_highlighted = review_highlighted.replace(token, Back.RED + token + Back.RESET)     # negatieve token rood markeren

print(review_highlighted)

nieuw concept in gent ,  maar dat kan volgens mij toch beter .  de meeste cornflakes waren gewoon de basic soorten .  ook wat duur voor de hoeveelheid die je krijgt ,  vooral met de toppings zijn ze zuinig .  en als je ontbijt aanbiedt ,  geef de mensen dan toch ook wat meer keuze voor hun koffie . 


<div style='color: #690027;' markdown="1">
    <h2>7. Oefening</h2> 
</div>

Bekijk de volgende reviews. 

-  Kies er een uit.
-  Welk sentiment koppel jij er intuïtief aan?
-  Leidt het lexicon tot hetzelfde resultaat? 

> Sunrisewater is net wat België nodig heeft. Met het obesitasprobleem dat toch wel aan een opmars bezig is, kunnen we alle initiatieven gebruiken om de jeugd weer gewoon water te laten drinken in plaats van die Amerikaanse bucht! Het smaakt geweldig en wat nog beter is, is dat je het gewoon op elke straathoek kan vinden! Echt geweldig! Vooral de pink and yellow is ten zeerste aan te raden.

>  Salé & Sucré staat bekend voor zijn super lekkere en originele cocktails, helaas was er geen alcoholvrije variant te verkrijgen. Onze BOB van dienst moest het dan maar bij frisdrank houden.

>  Het was superleuk om eens te mogen proeven van de Filipijnse keuken. De gerechten zaten goed in elkaar, de porties waren zeker groot genoeg en de smaken zaten helemaal goed. Voor herhaling vatbaar!

>  Gezellige sfeer, lekkere koffie en een mooi interieur. De combinatie van een studiebar en een babbelbar is een geniaal idee! Studeren met een lekker bakkie koffie, een overheerlijk hapje en samen met andere studenten, werkt enorm motiverend. Het interieur is enorm rustgevend met weinig afleiding, waardoor ik nog nooit zoveel heb kunnen doen!

> Wow, wat een coole restaurants! En het eten is er megalekker.

<div class="alert alert-box alert-success">
Proficiat, je hebt geleerd hoe een regelgebaseerd systeem voor sentimentanalyse werkt!
</div>

<img src="images/cclic.png" alt="Banner" align="left" style="width:100px;"/><br><br>
Notebook Chatbot, zie <a href="http://www.aiopschool.be">AI Op School</a>, van C. Van Hee, V. Hoste, F. wyffels, T. Neutens, Z. Van de Staey & N. Gesquière is in licentie gegeven volgens een <a href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Naamsvermelding-NietCommercieel-GelijkDelen 4.0 Internationaal-licentie</a>. 